core/drivers/stm32_etzpc.c

```diff
@@ -66,263 +61,400 @@
+	struct stm32_etzpc_driver_data *ddata = dev->ddata;
+	vaddr_t base = dev->pdata.base.va;
+	uint32_t reg = io_read32(base + ETZPC_HWCFGR);
 
-	hwcfg->num_tzma = (reg & ETZPC_HWCFGR_NUM_TZMA_MASK) >>
+	ddata->num_tzma = (reg & ETZPC_HWCFGR_NUM_TZMA_MASK) >>
 			  ETZPC_HWCFGR_NUM_TZMA_SHIFT;
-	hwcfg->num_per_sec = (reg & ETZPC_HWCFGR_NUM_PER_SEC_MASK) >>
+	ddata->num_per_sec = (reg & ETZPC_HWCFGR_NUM_PER_SEC_MASK) >>
 			     ETZPC_HWCFGR_NUM_PER_SEC_SHIFT;
-	hwcfg->num_ahb_sec = (reg & ETZPC_HWCFGR_NUM_AHB_SEC_MASK) >>
+	ddata->num_ahb_sec = (reg & ETZPC_HWCFGR_NUM_AHB_SEC_MASK) >>
 			     ETZPC_HWCFGR_NUM_AHB_SEC_SHIFT;
-	hwcfg->chunk_size = (reg & ETZPC_HWCFGR_CHUNCKS1N4_MASK) >>
-			    ETZPC_HWCFGR_CHUNCKS1N4_SHIFT;
+
+	DMSG("ETZPC revision 0x%02"PRIx8", per_sec %u, ahb_sec %u, tzma %u",
+	     io_read8(base + ETZPC_VERR),
+	     ddata->num_per_sec, ddata->num_ahb_sec, ddata->num_tzma);
 }
 
-static void init_device_from_hw_config(struct etzpc_instance *dev,
-				       paddr_t pbase)
+static void fdt_etzpc_conf_decprot(struct etzpc_device *dev,
+				   const void *fdt, int node)
 {
-	struct etzpc_hwcfg hwcfg = { };
+	const fdt32_t *cuint = NULL;
+	size_t i = 0;
+	int len = 0;
+
+	cuint = fdt_getprop(fdt, node, "default-access-conf", &len);
+	if (!cuint) {
+		DMSG("No ETZPC DECPROT configuration in DT");
+		return;
+	}
+
+	clk_enable(dev->pdata.clk);
 
-	assert(!dev->base.pa && cpu_mmu_enabled());
-	dev->base.pa = pbase;
-	dev->base.va = (vaddr_t)phys_to_virt(dev->base.pa, MEM_AREA_IO_SEC, 1);
-	assert(etzpc_base());
+	for (i = 0; i < len / sizeof(uint32_t); i += 2) {
+		uint32_t value = fdt32_to_cpu(cuint[i + 1]);
+		uint32_t id = value & ETZPC_ID_MASK;
+		uint32_t mode = (value & ETZPC_MODE_MASK) >> ETZPC_MODE_SHIFT;
+		bool lock = value & ETZPC_LOCK_MASK;
+		enum etzpc_decprot_attributes attr = ETZPC_DECPROT_MAX;
 
-	get_hwcfg(&hwcfg);
-	dev->num_tzma = hwcfg.num_tzma;
-	dev->num_per_sec = hwcfg.num_per_sec;
-	dev->num_ahb_sec = hwcfg.num_ahb_sec;
+		if (!valid_decprot_id(dev, id)) {
+			DMSG("Invalid DECPROT %"PRIu32, id);
+			panic();
+		}
 
-	DMSG("ETZPC revison 0x02%" PRIu8 ", per_sec %u, ahb_sec %u, tzma %u",
-	     io_read8(etzpc_base() + ETZPC_VERR),
-	     hwcfg.num_per_sec, hwcfg.num_ahb_sec, hwcfg.num_tzma);
+		attr = etzpc_binding2decprot(mode);
+		etzpc_do_configure_decprot(dev, id, attr);
 
-	init_pm(dev);
+		if (lock)
+			etzpc_do_lock_decprot(dev, id);
+	}
+
+	clk_disable(dev->pdata.clk);
+}
+
+static TEE_Result init_etzpc_from_dt(struct etzpc_device *etzpc_dev,
+				     const void *fdt, int node)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct dt_node_info etzpc_info = { };
+	struct io_pa_va base = { };
+	int len = 0;
+
+	fdt_fill_device_info(fdt, &etzpc_info, node);
+	if (etzpc_info.reg == DT_INFO_INVALID_REG ||
+	    etzpc_info.reg_size == DT_INFO_INVALID_REG_SIZE)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	base.pa = etzpc_info.reg;
+	etzpc_dev->pdata.name = fdt_get_name(fdt, node, &len);
```

Comment:
should `strdup()` the name to not reference the DT once inits are completed.
Note to also free `etzpc_dev->pdata.name` in `stm32_etzpc_free()`.